### PR TITLE
Ignore patches from workspace members

### LINKED ISSUE #4448: Consider simplify workspace - package `[patch]` table relation
With #4086, workspaces will have `[patch]` table support and the current relation between workspace and package `[patch]` table is the following:

'If the workspace manifest file contains a `[patch]` table, a member package's manifest file cannot contain any `[patch]` declarations.'

@mitchmindtree stated rust's approach is a little bit simpler which is (https://github.com/FuelLabs/sway/pull/4086/files#r1163520745):

>  i.e. "If workspace exists, patch table must go in workspace manifest. Otherwise patch table can go in package manifest."

### PR DESCRIPTION
**Pull request by cr-fuel**
## Description

Take the cargo approach of ignoring patches of workspace members, printing a warning, and only accepting patches defined at the workspace root file.

Fixes #4448

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by cr-fuel** [comment]
Here is the output of the warning

```
➜  patch_workspace_and_package git:(cr/simplify-workspace-package) ✗ ~/projects/fuel/sway/target/debug/forc build
warning: patch for the non root package will be ignored, specify patch at the workspace root:
warning: package: /Users/cr-fuel/projects/fuel/sway/forc-pkg/tests/invalid/patch_workspace_and_package/Forc.toml
  Fetching std git+https://github.com/fuellabs/sway?tag=v0.45.0#92dc9f361a9508a940c0d0708130f26fa044f6b3
  Creating a new `Forc.lock` file. (Cause: lock file did not exist)
    Adding core
    Adding std git+https://github.com/fuellabs/sway?tag=v0.45.0#92dc9f361a9508a940c0d0708130f26fa044f6b3
   Created new lock file at /Users/cr-fuel/projects/fuel/sway/forc-pkg/tests/invalid/patch_workspace_and_package/Forc.lock
 Compiling library core (/Users/cr-fuel/.forc/git/checkouts/std-9be0d6062747ea7/92dc9f361a9508a940c0d0708130f26fa044f6b3/sway-lib-core)
 Compiling library std (git+https://github.com/fuellabs/sway?tag=v0.45.0#92dc9f361a9508a940c0d0708130f26fa044f6b3)
 Compiling contract test_package (/Users/cr-fuel/projects/fuel/sway/forc-pkg/tests/invalid/patch_workspace_and_package/test_package)
  Finished debug in 9.584755575s
  contract test_package
      Bytecode size: 68 bytes
```

File: forc-pkg/src/manifest.rs [hunk 1]
```diff
@@ -285,19 +285,29 @@ impl PackageManifestFile {
...
-            .ok()
-            .flatten()
-            .and_then(|workspace| workspace.patch.clone());
-        let package_patches = self.patch.clone();
-        match (workspace_patches, package_patches) {
-            (Some(_), Some(_)) => bail!("Found [patch] table both in workspace and member package's manifest file. Consider removing [patch] table from package's manifest file."),
-            (Some(workspace_patches), None) => Ok(workspace_patches.into_iter()),
-            (None, Some(pkg_patches)) => Ok(pkg_patches.into_iter()),
-            (None, None) => Ok(BTreeMap::default().into_iter()),
+        if let Some(workspace) = self.workspace().ok().flatten() {
```
Comment (JoshuaBatty) [inline forc-pkg/src/manifest.rs:291]:
```suggestion
       if let Some(workspace) = self.workspace().ok().flatten() {
            // If workspace is defined, passing a local patch is a warning, but the global patch is used
            if self.patch.is_some() {
                println_warning(
                    "patch for the non root package will be ignored, specify patch at the workspace root:"
                );
                println_warning(&format!(
                    "package: {}",
                    workspace.path().to_str().unwrap_or_default()
                ));
            }
            return Ok(workspace.patch.as_ref().cloned().unwrap_or_default().into_iter());
        }
        Ok(self.patch.as_ref().cloned().unwrap_or_default().into_iter())
```

I think this is a little cleaner.

File: forc-pkg/src/manifest.rs [hunk 2]
```diff
@@ -285,19 +285,28 @@ impl PackageManifestFile {
...
+            // If workspace is defined, passing a local patch is a warning, but the global patch is used
+            if self.patch.is_some() {
+                println_warning(
+                    "patch for the non root package will be ignored, specify patch at the workspace root:"
```
Comment (sdankel) [inline forc-pkg/src/manifest.rs:295]: nit:

```suggestion
                    "Patch for the non root package will be ignored."
```

File: forc-pkg/src/manifest.rs [hunk 3]
```diff
@@ -285,19 +285,28 @@ impl PackageManifestFile {
...
+                );
+                println_warning(&format!(
+                    "package: {}",
+                    workspace.path().to_str().unwrap_or_default()
```
Comment (sdankel) [inline forc-pkg/src/manifest.rs:299]: Should we have a different warning in case workspace.path() is empty?

File: forc-pkg/src/manifest.rs [hunk 4]
```diff
@@ -285,19 +285,28 @@ impl PackageManifestFile {
...
-            (None, Some(pkg_patches)) => Ok(pkg_patches.into_iter()),
-            (None, None) => Ok(BTreeMap::default().into_iter()),
+        if let Some(workspace) = self.workspace().ok().flatten() {
+            // If workspace is defined, passing a local patch is a warning, but the global patch is used
+            if self.patch.is_some() {
+                println_warning(
+                    "patch for the non root package will be ignored, specify patch at the workspace root:"
+                );
+                println_warning(&format!(
+                    "package: {}",
```
Comment (sdankel) [inline forc-pkg/src/manifest.rs:298]:
```suggestion
                    "Specify patch at the workspace root: {}",
```

**Review by JoshuaBatty — DISMISSED** [review]
LGTM, good work.

File: forc-pkg/tests/invalid/patch_workspace_and_package/test_package/Forc.toml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "test_package"
+
+[dependencies]
+
+[patch.'https://github.com/fuellabs/sway']
+std = { git = "https://github.com/fuellabs/sway", branch = "test" }
```
Comment (sdankel) [inline forc-pkg/tests/invalid/patch_workspace_and_package/test_package/Forc.toml:10]: Will the tests using this fail if the branch `test` gets deleted?

File: forc-pkg/src/manifest.rs (hunk 3 above)
Comment (cr-fuel) [reply forc-pkg/src/manifest.rs:299]: If we want to copy what cargo does, we shouldn't. Packages inside workspaces are not allowed to have their own patch, even if no patch is provided into the namespace.

File: forc-pkg/tests/invalid/patch_workspace_and_package/test_package/Forc.toml (hunk 5 above)
Comment (cr-fuel) [reply forc-pkg/tests/invalid/patch_workspace_and_package/test_package/Forc.toml:10]: The test will just parse the toml files and request the list of patches. No further action will be performed.

File: forc-pkg/src/manifest.rs (hunk 3 above)
Comment (cr-fuel) [reply forc-pkg/src/manifest.rs:297]: What is your suggestion @sdankel? Do we take a different path /cc @JoshuaBatty

**Review by JoshuaBatty — APPROVED** [review]

**Pull request merged**